src/optimization.py

```diff
@@ -0,0 +1,363 @@
+        y = to_numpy(self.objective['y'])
+        GhAb = self.constraints.to_GhAb()
+        N = X.shape[1]
+        T = X.shape[0]
+
+        # Initiate an empty model
+        self.model = gp.Model('portfolio')
+        self.model.Params.LogToConsole = 0
+
+        # Add matrix variable for the asset weights
+        lb = to_numpy(self.constraints.box['lower'])
+        ub = to_numpy(self.constraints.box['upper'])
+        x = self.model.addMVar(N, lb=lb, ub=ub, vtype=gp.GRB.CONTINUOUS, name='x')
+
+        # Auxiliary variables to deal with the abs() function
+        aux_lad_pos = self.model.addMVar(T, lb=np.zeros(T), name='aux_lad_pos')
+        aux_lad_neg = self.model.addMVar(T, lb=np.zeros(T), name='aux_lad_neg')
+        self.model.addConstr(X @ x + aux_lad_pos - aux_lad_neg == y, name='aux_lad')
+
+        # Objective function
+        self.model.setObjective(aux_lad_pos.sum() + aux_lad_neg.sum(), gp.GRB.MINIMIZE)
+
+        # Add linear inequality constraints
+        if GhAb['G'] is not None:
+            self.model.addConstr(GhAb['G'] @ x <= GhAb['h'], name='Gh')
+
+        # Add linear equality constraints
+        if GhAb['A'] is not None:
+            self.model.addConstr(GhAb['A'] @ x == GhAb['b'], name='Ab')
+
+        # Add quadratic inequality constraints
+        quadcon = self.constraints.quadratic
+        if quadcon:
+            for key, value in quadcon.items():
+                if isinstance(value, dict):
+                    self.model.addConstr(value['q'].T @ x + (x @ value['Qc'] @ x) <= value['rhs'], key)
+
+        # Leverage constraint
+        if 'leverage' in self.constraints.l1.keys():
+            levcon = self.constraints.l1['leverage']
+            # Auxiliary variables to deal with the abs() function
+            aux_lvrg_pos = self.model.addMVar(N, lb=np.zeros(N), name='aux_lvrg_pos')
+            aux_lvrg_neg = self.model.addMVar(N, lb=np.zeros(N), name='aux_lvrg_neg')
+            self.model.addConstr(aux_lvrg_pos - aux_lvrg_neg == x, name='aux_leverage')
+            self.model.addConstr(aux_lvrg_pos.sum() + aux_lvrg_neg.sum() <= levcon['rhs'], 'leverage_budget')
+
+        # Solve and store results
+        self.model.optimize()
+        solved = True if self.model.status == 2 or (
+                    self.model.status == 13 and self.params['allow_suboptimal']) else False
+
+        if solved:
+            weights = pd.Series(self.model.x[0:N], index=self.constraints.selection)
+            obj_val = self.model.objVal
+        else:
+            weights = pd.Series(np.nan, index=self.constraints.selection)
+            obj_val = np.nan
+        self.results = {'weights': weights.to_dict(),
+                        'objective': obj_val,
+                        'status': self.model.status}
+        self.model.dispose()
+        return None
+
+    def solve_qpsolvers(self) -> None:
+        # Note: Should use an interior point linear solver instead of qpsolvers
```

Comment:
To use an ipm for lp to solve which problem?

Reply:
This comment is from Cyril's code of LAD: the objective function is penalized by the negative parts of absolute tracking error with benchmark.